src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+using Session = Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal class ItemManager : ForegroundThreadAffinitizedObject, IAsyncCompletionItemManager
+    {
+        private readonly CompletionHelper _completionHelper;
+
+        private const int MaxMRUSize = 10;
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+
+        internal ItemManager(IThreadingContext threadingContext) : base(threadingContext)
+        {
+            _completionHelper = new CompletionHelper(isCaseSensitive: true);
+        }
+
+        public Task<ImmutableArray<VSCompletionItem>> SortCompletionListAsync(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionInitialDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            AssertIsBackground();
+
+            SubscribeEvents(session);
+            return Task.FromResult(data.InitialList.OrderBy(i => i.SortText).ToImmutableArray());
+        }
+
+        public Task<AsyncCompletionData.FilteredCompletionModel> UpdateCompletionListAsync(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+            => Task.FromResult(UpdateCompletionList(session, data, cancellationToken));
+
+        private AsyncCompletionData.FilteredCompletionModel UpdateCompletionList(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            AssertIsBackground();
+
+            SubscribeEvents(session);
+            if (!session.Properties.TryGetProperty(CompletionSource.HasSuggestionItemOptions, out bool hasSuggestedItemOptions))
+            {
+                // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                // For now, the default hasSuggestedItemOptions is false. We can discuss if the opposite is required.
+                hasSuggestedItemOptions = false;
+            }
+
+            var filterText = session.ApplicableToSpan.GetText(data.Snapshot);
+            var reason = data.Trigger.Reason;
+
+            // We do not care about the character in the case. We care about the reason only.
+            var roslynTrigger = Helpers.GetRoslynTrigger(data.Trigger, data.Trigger.Character);
+
+            if (!session.Properties.TryGetProperty(CompletionSource.InitialTriggerKind, out CompletionTriggerKind initialRoslynTriggerKind))
+            {
+                initialRoslynTriggerKind = CompletionTriggerKind.Invoke;
+            }
```

Comment:
when does this happen?  why are we even involved here if we weren't there to store this data?